[3.8] bpo-42531: Teach importlib.resources.path to handle packages without __file__

wkschwartz opened the pull request:
Fixes [bpo-42531](https://bugs.python.org/issue42531) for Python 3.8.

The issue also applies to 3.7. If this PR looks like it'll be accepted, I can cherry-pick it to the 3.7 branch and submit a follow-up PR.

<!-- issue-number: [bpo-42531](https://bugs.python.org/issue42531) -->
https://bugs.python.org/issue42531
<!-- /issue-number -->

Automerge-Triggered-By: GH:jaraco

wkschwartz committed:
Fix bpo-42531: Teach importlib.resources.path to handle packages without __file__

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

jaraco commented on Lib/test/test_importlib/test_path.py:
```diff
@@ -26,6 +27,17 @@ def test_reading(self):
     data = data01
 
+    def test_package_spec_origin_is_None(self):
+        import pydoc_data
```
I'm a little uneasy depending on the implementation details (presence, name, contents) of a built-in module. In the `util` module, there's a routine that constructs a package based on parameters. Perhaps something like that should be used to build a contrived package that matches the desired interface.

jaraco commented on Lib/test/test_importlib/test_path.py:
```diff
@@ -26,6 +27,17 @@ def test_reading(self):
 class PathDiskTests(PathTests, unittest.TestCase):
     data = data01
 
+    def test_package_spec_origin_is_None(self):
```
Probably since this package represents a package _not_ on the file system (PathDisk), it should be in a separate test case. Does the test depend at all on PathTests or PathDiskTests.data?

jaraco commented on Lib/test/test_importlib/test_path.py:
```diff
@@ -26,6 +27,17 @@ def test_reading(self):
+        # Barge past path's sanity checks by ensuring spec.loader.is_resource
+        # returns False.
+        with swap_attr(spec, "origin", None), \
+            swap_attr(spec.loader, "is_resource", lambda *args: False), \
```
This part of the test feels a little brittle, relying on other implementation details. Another reason to build a fixture that represents the (previously-) unmet expectation.

wkschwartz committed:
Fix test per review comments

https://github.com/python/cpython/pull/23611#pullrequestreview-564816417

This creates a package in memory and sets its __spec__.origin to None manually.
Otherwise the test proceeds like the others PathTests subclasses.

jaraco committed:
Apply tweaks to decouple tests in test_importlib.test_path and apply black formatting.

jaraco reviewed (approved):
Looks great. Thanks for all the work on this.
